src/transaction/builder/mod.rs

```diff
@@ -0,0 +1,353 @@
+        tx_data.set_output(idx, output);
+        Ok(())
+    }
+}
+
+macro_rules! celloutput_capacity {
+    ($output:expr) => {{
+        let tmp_capacity: u64 = $output.capacity().unpack();
+        tmp_capacity
+    }};
+}
+
+impl CkbTransactionBuilder for SimpleTransactionBuilder {
+    fn build(
+        &mut self,
+        contexts: &HandlerContexts,
+    ) -> Result<TransactionWithScriptGroups, TxBuilderError> {
+        let mut lock_groups: HashMap<Byte32, ScriptGroup> = HashMap::default();
+        let mut type_groups: HashMap<Byte32, ScriptGroup> = HashMap::default();
+        let mut outputs_capacity = 0u64;
+
+        for (output_idx, output) in self.tx.get_outputs().clone().iter().enumerate() {
+            outputs_capacity += celloutput_capacity!(output);
+            if let Some(t) = &output.type_().to_opt() {
+                let script_group = type_groups
+                    .entry(t.calc_script_hash())
+                    .or_insert_with(|| ScriptGroup::from_type_script(t));
+                script_group.output_indices.push(output_idx);
+                Self::handle_script(&mut self.tx, &self.configuration, script_group, contexts)?;
+            }
+        }
+
+        let mut state = BalanceState::Init;
+        let mut inputs_capacity = 0u64;
+        let mut mini_change_capacity = 0u64;
+        let calculator = self.configuration.fee_calculator();
+        for (input_index, input) in
+            InputView::new(&self.transaction_inputs, &mut self.input_iter).enumerate()
+        {
+            let input = input?;
+            self.tx.input(input.cell_input());
+            let previous_output = input.previous_output();
+            self.tx.witness(packed::Bytes::default());
+            let lock_script = previous_output.lock();
+            let script_group = lock_groups
+                .entry(lock_script.calc_script_hash())
+                .or_insert_with(|| ScriptGroup::from_lock_script(&lock_script));
+            script_group.input_indices.push(input_index);
+            // add cellDeps and set witness placeholder
+            Self::handle_script(&mut self.tx, &self.configuration, script_group, contexts)?;
+
+            if let Some(t) = &previous_output.type_().to_opt() {
+                let script_group = type_groups
+                    .entry(t.calc_script_hash())
+                    .or_insert_with(|| ScriptGroup::from_type_script(t));
+
+                script_group.input_indices.push(input_index);
+                Self::handle_script(&mut self.tx, &self.configuration, script_group, contexts)?;
+            }
+            inputs_capacity += celloutput_capacity!(previous_output);
+            // check if there is enough capacity for output capacity and change
+            let fee = calculator.fee_with_tx_data(&self.tx);
+            let change_capacity =
+                (inputs_capacity + self.reward).checked_sub(outputs_capacity + fee);
+            if let Some(mut change_capacity) = change_capacity {
```

Comment:
1. change cell is not created at first place, it may not need change cell.
2. if has change capacity, 
     a. Need to put in change cell, then add change output, which will increase fee, so new_fee will need to be calculated.
          - if change can cover fee, it's balanced.
          - if change not enough, need to find another input cell for the change cell.
     b. Change capacity is small, and can be put to transaction fee, then it's balanced.
     c. Change capacity is small, and can be put to receiver, increase receiver cell's capacity.